transport/src/main/java/io/scalecube/transport/Transport.java

```diff
@@ -241,13 +256,25 @@ public void send(@CheckForNull TransportEndpoint endpoint, @CheckForNull Message
   }
 
   @Override
-  public void send(@CheckForNull TransportEndpoint endpoint, @CheckForNull Message message,
-      @Nullable SettableFuture<Void> promise) {
+  public void send(@CheckForNull final TransportEndpoint endpoint, @CheckForNull final Message message,
+      @Nullable final SettableFuture<Void> promise) {
     checkArgument(endpoint != null);
     checkArgument(message != null);
-    TransportChannel transportChannel = getOrConnect(endpoint.socketAddress());
     // TODO [AK]: check that channel endpoint id correspond to provided endpoint id; fail otherwise
-    transportChannel.send(message, promise);
+    Futures.addCallback(getOrConnect(endpoint.socketAddress()), new FutureCallback<TransportChannel>() {
```

Comment:
Don't register callback on any call. First check that future is completed and if yes, just call get and send; only otherwise register callback.